include/oboe/FullDuplexStream.h

```diff
@@ -0,0 +1,220 @@
+                if (!result) {
+                    // Ignore errors because input stream may not be started yet.
+                    break;
+                }
+                actualFramesRead = result.value();
+                totalFramesRead += actualFramesRead;
+            } while (actualFramesRead > 0);
+            // Only counts if we actually got some data.
+            if (totalFramesRead > 0) {
+                mCountCallbacksToDrain--;
+            }
+
+        } else if (mCountInputBurstsCushion > 0) {
+            // Let the input fill up a bit so we are not so close to the write pointer.
+            mCountInputBurstsCushion--;
+
+        } else if (mCountCallbacksToDiscard > 0) {
+            mCountCallbacksToDiscard--;
+            // Ignore. Allow the input to reach to equilibrium with the output.
+            ResultWithValue<int32_t> resultAvailable = getInputStream()->getAvailableFrames();
+            if (!resultAvailable) {
+                callbackResult = DataCallbackResult::Stop;
+            } else {
+                int32_t framesAvailable = resultAvailable.value();
+                if (framesAvailable >= mMinimumFramesBeforeRead) {
+                    ResultWithValue<int32_t> resultRead = readInput(numFrames);
+                    if (!resultRead) {
+                        callbackResult = DataCallbackResult::Stop;
+                    }
+                }
+            }
+        } else {
+            int32_t framesRead = 0;
+            ResultWithValue<int32_t> resultAvailable = getInputStream()->getAvailableFrames();
+            if (!resultAvailable) {
+                callbackResult = DataCallbackResult::Stop;
+            } else {
+                int32_t framesAvailable = resultAvailable.value();
+                if (framesAvailable >= mMinimumFramesBeforeRead) {
+                    // Read data into input buffer.
+                    ResultWithValue<int32_t> resultRead = readInput(numFrames);
+                    if (!resultRead) {
+                        callbackResult = DataCallbackResult::Stop;
+                    } else {
+                        framesRead = resultRead.value();
+                    }
+                }
+            }
+
+            if (callbackResult == DataCallbackResult::Continue) {
+                callbackResult = onBothStreamsReady(mInputBuffer.get(), framesRead,
+                                                    audioData, numFrames);
+            }
+        }
+
+        if (callbackResult == DataCallbackResult::Stop) {
+            getInputStream()->requestStop();
+        }
+
+        return callbackResult;
+    }
+
+    int32_t getMNumInputBurstsCushion() const {
+        return mNumInputBurstsCushion;
+    }
+
+    /**
+     * Number of bursts to leave in the input buffer as a cushion.
+     * Typically 0 for latency measurements
+     * or 1 for glitch tests.
+     *
+     * @param mNumInputBurstsCushion
+     */
+    void setMNumInputBurstsCushion(int32_t numBursts) {
```

Comment:
This is a cushion between the DSP and the application processor cursors to prevent collisions.